File: maxSumBST.c
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */

#include<stdio.h>
#include<stdlib.h>
#include<string.h>

void print_int_star(int *list, int len)
{
    for(int i=0;i<len;i++)
        printf("%d ",list[i]);
    putchar('\n');
}

void print_int_2star(int **matrix,int row,int col)
{
    for(int i=0;i<row;i++)
        print_int_star(matrix[i],col);
}

#define max(a,b) (((a)>(b))?(a):(b))

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

typedef struct {
    int max;
    int min;
    int maxSum;
    bool isBST;
} node;

int maxSumBST(struct TreeNode* root){
    if(root==NULL)
        return 0;
    return msb(root).maxSum;
}

node msb(struct TreeNode* root)
{
    if(root->left==NULL && root->right==NULL)
        return node(root->val,root->val,root->val,true);
    node out;
    out->isBST=true;
    if(root->left!=NULL)
    {
        node ol=msb(root->left);
        if(!ol.isBST)
            out.isBST=false;
        if(root.val<=ol.max)
            out.isBST=false;
        maxSum
    }
}
// not finished

    return out;
}

int main()
{

    return 0;
}
    

File: maxSumBST.py
from typing import Optional

from pytree import TreeNode


class Solution:
    def maxSumBST(self, root: Optional[TreeNode]) -> int:
        res = 0

        def find(node):
            # (s,right,ma,mi):

            if node is None:
                return (0, True, -123456, 123456)
            tl = find(node.left)
            tr = find(node.right)
            right = tl[1] and tr[1] and tl[2] < node.val and node.val < tr[3]
            s = tl[0] + tr[0] + node.val
            lmax = tr[3]
            ma = max(tl[2], tr[2], node.val)
            mi = min(tl[3], tr[3], node.val)
            # print(node.val,s,right,ma,mi)

            if right:
                nonlocal res
                res = max(res, s)

            return (s, right, ma, mi)

        find(root)

        return res


root = '[1,4,3,2,4,2,5,null,null,null,null,null,null,4,6]'
root = '[4,3,null,1,2]'
root = '[-4,-2,-5]'
root = '[2,1,3]'
root = '[5,4,8,3,null,6,3]'
root = TreeNode.fromStrList(root)
print(Solution().maxSumBST(root))
